### ISSUE TITLE
Change cert@rancher.dev and require users to explicitly set an email on rio install

### ISSUE THREAD

**Issue by rancher-max** [rancher/rio#889]
The ask came from Let's Encrypt:

> The cert@rancher.dev default is particularly problematic because it is not a deliverable address, and as far as I can tell rancher.dev is not actually owned by Rancher. Could I ask you to remove the default email address from Rio and require that users explicitly set an email before Rio creates an account?

We propose adding an (optional) flag for a user to supply a valid email address.
We may want to update docs as well to explain the purpose of the email flag and that if you want expiry notifications.

**Comment by vincent99** [comment]
FWIW there's no MX record but it is a domain we own.

**Comment by jsha** [comment]
Hi,

I hope you can fix this issue soon. At Let's Encrypt we're likely to start blocking new account creation from this address soon since it is undeliverable.

**Comment by vincent99** [comment]
@jsha We value the service you are providing and would like to keep you guys happy.  I don't know exactly what all your concerns are yet so I would like to here more about what the problem is from your perspective, but here are a few ideas that come to mind that we could do anyway.

1) I could add a MX record and make `cert@rancher.dev` "deliverable" today.  Since every user still shares the same email it's not clear that this accomplishes much for anybody (i.e. I'd just deliver them to /dev/null).  But if the bounces themselves are a problem on your side we can do this.

2) Is providing an email to you actually required?  https://letsencrypt.org/docs/expiration-emails/ says "**If** you provide an email".  If permitted, the next version of Rio could just not provide any email address to you at all.

3) Each Rio install gets a random 6-character ID which we create a DNS record for `XXXXXX.on-rio.io` and then request a `*.XXXXXX.on-rio.io` cert from you.  So we could submit the email to you as `XXXXXX@on-rio.io`, and make those all accept mail.  These would still be "delivered" to /dev/null, because Rio is open-source and runs on random user's hardware and we have no idea who they are.  But the net result would be that each installation/user gets its own email address instead of all sharing one, if that helps you.

4) We could add a flag or installation prompt for the user to provide their email.  This seems totally reasonable as an _option_, as I'm sure some users would actually want their expiration notices.  I would prefer to be able to fall back to _something_ deterministic (like one of the options above) rather than **requiring** that users provide an email though.  There's no practical way for us to validate them so if we require input instead of opting-in you will end up with a good fraction that are random garbage.  And apparently your side is also doing no validation, since you take our current value one that has no MX record...

**Comment by davidnuzik** [comment]
We discussed with the team and proposed with interactive installs prompting the user for an email and for non-interactive providing a flag that must be provided with the email. We still cannot validate the emails of course.

**Comment by jsha** [comment]
Good point, I sort of started in the middle. I should start from the beginning and explain why we care so much about this.

Sometimes clients go out of control and send us enormous amounts of junk traffic. Cert-manager in particular has had this problem very often. We've worked with Jetstack over the last couple of years to fix a series of bugs, and we recently started [blocking older cert-manager versions](https://community.letsencrypt.org/t/blocking-old-cert-manager-versions/98753) to ensure people are upgrading to the fixed versions.

When we see a client that is sending excessive traffic, we've found that just blocking the client doesn't really fix the problem, since it's almost always a client bug rather than an operator error. However, we don't have access to the client logs so we can't help the client authors debug. In that situation, we send email to the address on file for that client, asking them to collect logs and file a bug against the ACME client software so the issue can be fixed.

Recently, two of our top clients are Rancher clients ("go-acme/2 jetstack-cert-manager/v0.11.0-rio.1"). Each one is sending us a million requests per day, without ever issuing a certificate. I tried getting in touch with the operators of those clients by emailing the address on file (cert@rancher.dev), but got a bounce. In the absence of an email address, I have no way to reach the owners of these clients, and in turn I have no way to help you find the bug in Rancher and fix it. That leads to excess load on Let's Encrypt and increased costs.

You're correct; we don't require an email address to sign up. However, we strongly encourage it. And in particular we try to avoid the situation where someone winds up with no email, or a fake email, just because the software they were using or the setup guide they were following defaulted them into the no-email or fake-email situation.

Certbot handles this by requiring an email address by default, but offering a `--register-unsafely-without-email` flag to omit it. This clearly signals that providing an email is a happy path and omitting it is the risky path. We've had very good success with this. In my experience, most Certbot users provide email addresses, and we have been able to reach people about deprecations and needed upgrades.

So, for Rancher's case in particular we care about this because it's important for us to be able to reach out to people about their broken clients that are causing us problems. We also sometimes need to send emails about upcoming changes on our side, like the ACMEv1 deprecation, that might break renewals. And of course, the expiration emails. Even though Rancher renews automatically, there are a lot of different reasons it might fail. We want people to perceive Rancher and Let's Encrypt as reliable, which means that they should receive expiration emails *before* their site goes offline.

> We could add a flag or installation prompt for the user to provide their email. This seems totally reasonable as an option, as I'm sure some users would like to get their expiration notices. We renew it automatically if the Rio cluster is running, but this would be a good reminder before you lose the name forever if the cluster is down for a while.

My ideal would be something like the Certbot model: There's a flag to provide an email; if you don't provide an email, you have to additionally provide a `--register-unsafely-without-email` flag to acknowledge that that's not the happy path.

> I could add a MX record and make cert@rancher.dev "deliverable" today. Since every user still shares the same email it's not clear that this accomplishes much for anybody 

Good point, I probably over-emphasized the deliverability aspect here. We do care about the cost of sending unnecessary emails, and the effect on our bounce rate, but what we care about most is that we be able to reach users, if they are willing to provide their email address (which experience has shown is most of them, if client software encourages it).

**Comment by vincent99** [comment]
Ok I was already in the middle of adding the mailbox for it so mail to cert@rancher.dev should deliver now for whatever that's worth.

Will discuss with the rest of the team today but `--email` or e.g. `--yes-really-no-email` seems reasonable to me.

Besides the email issue we'd also like to not spam your API...  We can't force existing installs to change but is there a specific fix in cert-manager 0.11.1 or 0.12 that you think us upgrading would help?  Or do you think we're doing something that is causing it?  Rio is pretty new/beta and has nowhere near millions of active users so I wouldn't expect it to be an even noteworthy volume of traffic for you.

**Comment by vincent99** [comment]
So on `rio install`, a user will have to pass one of:

```
-m, --email                    Email address exclusively for Let's Encrypt to provide important notifications, certificate expiration warnings, and lost account recovery.  Never shared with Rancher.
--no-email                     Use Let's Encrypt with no email address.  This is strongly discouraged and lack of notifications may cause you to lose access to your certificates.
--disable-features=letsencrypt Do not use Let's Encrypt at all.
```

And save the email into a setting that the UI can then show/edit.

**Comment by jsha** [comment]
That sounds great! Thanks for working on this. BTW you can leave out "lost account recovery." That's still in Certbot from the early days when we thought we might implement account recovery, but it never happened and is unlikely to. I should get around to editing the Certbot output.

**Comment by jsha** [comment]
> is there a specific fix in cert-manager 0.11.1 or 0.12 that you think us upgrading would help? Or do you think we're doing something that is causing it?

I followed up by email, but posting here for the sake of continuity: Two of our top ten clients by traffic volume are Rio clients. They follow a similar pattern we've seen in some older cert-manager clients: They are creating accounts over and over again, at a rate of a million API requests per day. They are mostly getting rate limited, but it still takes some resources to reject all those requests, and we'd rather get the issue solved before Rancher / Rio has tons of users. Unfortunately it's kind of hard to get to the root of because we're not in touch with the people whose clients are exhibiting the bug.

**Comment by rancher-max** [comment]
The email flags have been added and validated. Moving forward, all new users and those who upgrade their version of rio will have to either supply a flag to specify `--email` or `--no-email`, or they will be prompted to provide an email (or not: See https://github.com/rancher/rio/pull/931/files#diff-2430bedc5dc6765976857e57e16ddc16R217-R219 for the prompt specifics).

If no email is provided and Let's Encrypt is enabled, then we do currently still use `cert@rancher.dev` due to cert-manager doing validation on the email field.
